Add introduction to the section on variants and inductive types in the reference manual (#17800)
coq/coq: doc/sphinx/language/core/variants.rst

Variants
--------

The :cmd:`Variant` command allows defining types by listing
the :term:`inhabitants` of the type. This includes enumerated types from
programming languages, such as Booleans, characters or even the
degenerate cases of the unit and empty types. Variant types more
generally include enumerated types with arguments or even enumerated
types with parametric arguments such as option types, sum types,
etc. It also includes predicates or type families defined by cases
such as the Boolean reflection or equality predicates. Observing the
form of the term:`inhabitants` of a variant type is done by case analysis
using the :cmd:`match` expression. Each case is described by what is
called a :gdef:`constructor` of the type. For instance, Booleans have two
constructors: :g:`true` and :g:`false`.

When a constructor of a type takes an argument of that same type,
the type becomes recursive, in which case it can be either
:cmd:`Inductive` or :cmd:`CoInductive`. The keyword :cmd:`Variant`
is reserved for non-recursive types. Natural numbers, lists or streams cannot
be defined using :cmd:`Variant`.

> .. cmd:: Variant @ident {? @cumul_univ_decl } {* @binder } {? %| {* @binder } } {? : @type } := {? %| } {+| @constructor } {? @decl_notations }

Reviewer: There are 15 uses of `ident_decl` in the refman. If you think that it creates confusion and should be unfolded, we should do it for all the occurrences (at the risk of making grammar description longer). You can test this change by adding `ident_decl` to the SPLICE list of `common.edit_mlg` (currently starting on line 2260) and running `make doc_gram_rsts`. This change which doesn't match what is in `common.edit_mlg` / `orderedGrammar` is the reason for the linter failure.

Author: I was needing a way to refer to the name "ident" of the variant type. I followed the example of the Inductive section where ident_decl was already split. I have to particular preference otherwise.

Reviewer: It is split in `Inductive` because it uses `cumul_univ_decl` which is not used anywhere else. Typically the way we refer to production names that do not appear directly is to write "the `ident` (in `ident_decl`)".

Author: I see. Done. Hopefully as expected. If not, I can still change.

Variants
--------

The :cmd:`Variant` command allows defining types by listing

> the :term:`inhabitants` of the type. This includes enumerated types from

Reviewer: This is the reason why the compilation currently fails:

```suggestion
the :term:`inhabitants <inhabitant>` of the type. This includes enumerated types from
```

> The :cmd:`Variant` command allows defining types by listing
> the :term:`inhabitants <inhabitant>` of the type. This includes enumerated types from
> programming languages, such as Booleans, characters or even the

Reviewer 2: ```suggestion
The :cmd:`Variant` command allows defining types by listing
the :term:`inhabitants <inhabitant>` of the type.  Each inhabitant is
specified by a :gdef:`constructor`.  For instance, Booleans have two
constructors: :g:`true` and :g:`false`. Types can include enumerated types from
programming languages, such as Booleans, characters or even the
```

Moved two sentences up from below and simplified wording for constructors.  (not sure why the diff shown is not more granular today)

Author: Yes, much better.

degenerate cases of the unit and empty types. Variant types more
generally include enumerated types with arguments or even enumerated
types with parametric arguments such as option types, sum types,
etc. It also includes predicates or type families defined by cases
such as the Boolean reflection or equality predicates. Observing the

> form of the term:`inhabitants` of a variant type is done by case analysis

Reviewer: ```suggestion
form of the :term:`inhabitants <inhabitant>` of a variant type is done by case analysis
```

> using the :cmd:`match` expression. Each case is described by what is

Reviewer: `match` is not a command. This is what makes the current build fail:

```suggestion
using the `match` expression. Each case is described by what is
```

> using the `match` expression. Each case is described by what is
> called a :gdef:`constructor` of the type. For instance, Booleans have two
> constructors: :g:`true` and :g:`false`.

Reviewer 2: ```suggestion
using the `match` expression.
```

> The :cmd:`Variant` command allows defining types by listing
> the :term:`inhabitants <inhabitant>` of the type. This includes enumerated types from
> programming languages, such as Booleans, characters or even the
> degenerate cases of the unit and empty types. Variant types more
> generally include enumerated types with arguments or even enumerated
> types with parametric arguments such as option types, sum types,
> etc. It also includes predicates or type families defined by cases
> such as the Boolean reflection or equality predicates. Observing the
> form of the :term:`inhabitants <inhabitant>` of a variant type is done by case analysis
> using the :cmd:`match` expression. Each case is described by what is
> called a :gdef:`constructor` of the type. For instance, Booleans have two
> constructors: :g:`true` and :g:`false`.

Reviewer 2: ```suggestion
The :cmd:`Variant` command allows defining types by listing
the :term:`inhabitants <inhabitant>` of the type.  This includes enumerated types from
programming languages, such as Booleans, characters or even the
degenerate cases of the unit and empty types. Variant types more
generally include enumerated types with arguments or even enumerated
types with parametric arguments such as option types and sum types.
It also includes predicates or type families defined by cases
such as the Boolean reflection or equality predicates. Observing the
form of the :term:`inhabitants <inhabitant>` of a variant type is done by case analysis
using the :cmd:`match` expression. Each case is described by what is
called a :term:`constructor` of the type. For instance, Booleans have two
constructors: :g:`true` and :g:`false`.
```

Author: Adopting "etc" part.

For constructor, don't we need `gdef` to define it?

Reviewer 2: Yes, `gdef`.  [Homer nods :-)](https://en.wiktionary.org/wiki/even_Homer_nods#English)

> constructors: :g:`true` and :g:`false`.

Reviewer 2: Constructors apply equally to Variants, Inductives and CoInductives, so I think they should be described at the beginning of inductive.rst.

Probably should say constructors have names.

Author: Do you see a particular reason to define it at the beginning of inductive.rst rather than at the beginning of variants.rst (which comes first)?

Reviewer 2: It will be fine in variants.rst.  (I looked at the order in which the changes appeared in github rather than in the actual manual.  I just got back to California Tuesday evening, so I'm a bit more addled than usual...)

When constructors recursively take arguments in the type being
defined, the type becomes recursive, in which case it can be either
:cmd:`Inductive` or :cmd:`CoInductive`. The terminology :cmd:`Variant`
is thus specifically used to indicate a non-recursive type defined by

> constructors (for instance, natural numbers, lists or streams cannot

Reviewer 2: "stream" (an infinite sequence of numbers?) may not be familiar to your average software engineer.  I didn't know it.

Author: In Coq, streams are the following type:
```coq
CoInductive Stream (A : Type) := { hd : A ; tl : Stream A }.
```
This indeed generalizes infinite sequence of numbers to infinite sequence of any kind of objects.

Then, [coinductive types](https://en.wikipedia.org/wiki/Coinduction#Codata) is a generalization of streams.

> be defined using :cmd:`Variant`).

Author: I would prefer:
```
:cmd:`Inductive` or :cmd:`CoInductive`. The keyword :cmd:`Variant`
is reserved for non-recursive types.  Natural numbers, lists and streams cannot
be defined using :cmd:`Variant`.
```
if ok for you.

Variants
--------

> The :cmd:`Variant` command allows to define types by case on the inhabitants of the type. This includes enumerated types from programming languages, such as Booleans, characters or even the degenerate cases of the unit and empty types. Variant types more generally include enumerated types with arguments or even enumerated types with parametric arguments such as option types, sum types, etc. It also includes predicates or type families defined by cases such as the Boolean reflection or equality predicates. Observing the form of the inhabitant of a variant type is done by case analysis using the :cmd:`match` expression. Each case is described by what is called a :g:`constructor` of the type. For instance, Booleans have two constructors: :g:`true` and :g:`false`.

Reviewer 2: ```suggestion
The :cmd:`Variant` command allows defining types by case on the inhabitants of the type. This includes enumerated types from programming languages, such as Booleans, characters or even the degenerate cases of the unit and empty types. Variant types more generally include enumerated types with arguments or even enumerated types with parametric arguments such as option types, sum types, etc. It also includes predicates or type families defined by cases such as the Boolean reflection or equality predicates. Observing the form of the inhabitant of a variant type is done by case analysis using the :cmd:`match` expression. Each case is described by what is called a :g:`constructor` of the type. For instance, Booleans have two constructors: :g:`true` and :g:`false`.
```

Variant types are not recursive (your examples boolean and character are not)?
Maybe add "inhabitant" to the glossary if it's essential to use that word here.
"Constructor" should be in the glossary and defined more explicitly and less tersely.

Please split into multiple lines.

   .. exn:: The @natural th argument of @ident must be @ident in @type.
      :undocumented:

.. example::

  The Booleans, the unit type and the empty types are respectively defined by:

   .. coqtop:: in

      Variant bool : Type := true : bool | false : bool.
      Variant unit : Type := tt : unit.
      Variant Empty_set : Type :=.

  The option and sum types are defined by:

   .. coqtop:: in

      Variant option (A : Type) : Type := None : option A | Some : A -> option A.
      Variant sum (A B : Type) : Type := inl : A -> sum A B | inr : B -> sum A B.

  The Boolean reflection and equality predicates are defined by:

  .. coqtop:: in

     Variant reflect (P : Prop) : bool -> Set :=
     | ReflectT : P -> reflect P true
     | ReflectF : ~ P -> reflect P false.

>      Variant eq (A : Type) (x : A) : A -> Prop :=  eq_refl : eq x x.

Reviewer 3: Do we already know about implicit argument at this point? Otherwise maybe
```suggestion
     Variant eq (A : Type) (x : A) : A -> Prop :=  eq_refl : @eq A x x.
```
could help?
Would those two type require a bit more explanation?

Reviewer 4: That change still needs knowledge of implicit arguments to explan the @ 
Maybe disable automatic implicits instead

   .. exn:: The @natural th argument of @ident must be @ident in @type.
      :undocumented:

.. example::

  The Booleans, the unit type and the empty types are respectively defined by:

   .. coqtop:: in

      Variant bool : Type := true : bool | false : bool.
      Variant unit : Type := tt : unit.
      Variant Empty_set : Type :=.

  The option and sum types are defined by:

   .. coqtop:: in

      Variant option (A : Type) : Type := None : option A | Some : A -> option A.
      Variant sum (A B : Type) : Type := inl : A -> sum A B | inr : B -> sum A B.

>   *Boolean reflection* is a relation reflecting under the form of a

Reviewer 2: FWIW, "reflection" is probably worth adding to the glossary at some point.

Author: It is somehow more specfic to the library than to the logic.

  Boolean value when a given proposition :n:`P` holds. It can be
  defined as a two-constructor type family over :g:`bool`
  parameterized by the proposition :n:`P`:

  .. coqtop:: in

     Variant reflect (P : Prop) : bool -> Set :=
     | ReflectT : P -> reflect P true
     | ReflectF : ~ P -> reflect P false.

>   :term:`Leibniz Equality` is another example of variant type.

Reviewer 2: Actually, it's defined as a Inductive: `Inductive eq (A:Type) (x:A) : A -> Prop :=
    eq_refl : x = x :>A`

Reviewer: The actual keyword that is used `Inductive` or `Variant` does not matter. This is a variant because it is not recursive. Maybe we should make this clearer in the doc...

Author: We are facing a dilemma here. It is indeed artificial to use `Inductive` for a type which is neither inductive nor coinductive but `Variant` does not define `eq_ind`/`eq_rect`, which are pervasively used.

So, we would need to add a "Set Nonrecursive Elimination Schemes" beforehand. But why not, if it helps to make thinks clearer.

Reviewer 2: Probably worth mentioning the use of Inductive to create XX_ind and XX_rect as an alternative in the description of Variant.

   .. exn:: The @natural th argument of @ident must be @ident in @type.
      :undocumented:

.. example::

  The Booleans, the unit type and the empty types are respectively defined by:

   .. coqtop:: none

      Module FreshNameSpace.

   .. coqtop:: in

      Variant bool : Type := true : bool | false : bool.
      Variant unit : Type := tt : unit.
      Variant Empty_set : Type :=.

  The option and sum types are defined by:

   .. coqtop:: in

      Variant option (A : Type) : Type := None : option A | Some : A -> option A.
      Variant sum (A B : Type) : Type := inl : A -> sum A B | inr : B -> sum A B.

  *Boolean reflection* is a relation reflecting under the form of a
  Boolean value when a given proposition :n:`P` holds. It can be
  defined as a two-constructor type family over :g:`bool`
  parameterized by the proposition :n:`P`:

  .. coqtop:: in

     Variant reflect (P : Prop) : bool -> Set :=
     | ReflectT : P -> reflect P true
     | ReflectF : ~ P -> reflect P false.

  .. coqtop:: none

     End FreshNameSpace.

  :term:`Leibniz Equality` is another example of variant type.

.. note::

>    It is common in the standard library to use :cmd:`Inductive` in
>    place of :cmd:`Variant` even for non-recursive types for the only
>    purpose of automatically deriving associated schemes of name
>    :n:`@ident`\ ``_rect``, :n:`@ident`\ ``_ind``, :n:`@ident`\
>    ``_rec`` and :n:`@ident`\ ``_sind``.

Reviewer 2: ```suggestion
   The standard library commonly uses :cmd:`Inductive` in
   place of :cmd:`Variant` even for non-recursive types in order to
   automatically derive the schemes
   :n:`@ident`\ ``_rect``, :n:`@ident`\ ``_ind``, :n:`@ident`\
   ``_rec`` and :n:`@ident`\ ``_sind``.  (These schemes are also created
   for :cmd:`Variant` if the :flag:`Nonrecursive Elimination Schemes` flag is set.)
```

   .. exn:: The @natural th argument of @ident must be @ident in @type.
      :undocumented:

.. example::

  The Booleans, the unit type and the empty types are respectively defined by:

   .. coqtop:: none

      Module FreshNameSpace.

   .. coqtop:: in

      Variant bool : Type := true : bool | false : bool.
      Variant unit : Type := tt : unit.
      Variant Empty_set : Type :=.

  The option and sum types are defined by:

   .. coqtop:: in

      Variant option (A : Type) : Type := None : option A | Some : A -> option A.
      Variant sum (A B : Type) : Type := inl : A -> sum A B | inr : B -> sum A B.

  *Boolean reflection* is a relation reflecting under the form of a
  Boolean value when a given proposition :n:`P` holds. It can be
  defined as a two-constructor type family over :g:`bool`
  parameterized by the proposition :n:`P`:

  .. coqtop:: in

     Variant reflect (P : Prop) : bool -> Set :=
     | ReflectT : P -> reflect P true
     | ReflectF : ~ P -> reflect P false.

  .. coqtop:: none

     End FreshNameSpace.

>   :term:`Leibniz Equality` is another example of variant type.

Reviewer: Casing matters (this is making the build fail):

```suggestion
  :term:`Leibniz equality` is another example of variant type.
```
